temporal-sdk/src/main/java/io/temporal/internal/worker/LocalActivityWorker.java

```diff
@@ -256,51 +263,70 @@ public boolean dispatch(
     private boolean submitANewExecution(
         @Nonnull LocalActivityExecutionContext executionContext,
         @Nonnull PollActivityTaskQueueResponse.Builder activityTask,
-        @Nullable Deadline acceptanceDeadline) {
+        @Nullable Deadline heartbeatDeadline) {
+      long acceptanceTimeoutMs = 0;
+      boolean timeoutIsScheduleToStart = false;
+      if (heartbeatDeadline != null) {
+        acceptanceTimeoutMs = heartbeatDeadline.timeRemaining(TimeUnit.MILLISECONDS);
+      }
+      Duration scheduleToStartTimeout = executionContext.getScheduleToStartTimeout();
+      if (scheduleToStartTimeout != null) {
+        long scheduleToStartTimeoutMs = scheduleToStartTimeout.toMillis();
+        if (scheduleToStartTimeoutMs > 0 && scheduleToStartTimeoutMs < acceptanceTimeoutMs) {
+          acceptanceTimeoutMs = scheduleToStartTimeoutMs;
+          timeoutIsScheduleToStart = true;
+        }
+      }
       try {
-        boolean accepted;
-        if (acceptanceDeadline == null) {
-          newExecutionsBackpressureSemaphore.acquire();
-          accepted = true;
+        SlotPermit permit = null;
+        SlotReservationData reservationCtx =
+            new SlotReservationData(taskQueue, options.getIdentity(), options.getBuildId());
+        if (acceptanceTimeoutMs <= 0) {
+          permit = slotSupplier.reserveSlot(reservationCtx);
         } else {
-          long acceptanceTimeoutMs = acceptanceDeadline.timeRemaining(TimeUnit.MILLISECONDS);
-          if (acceptanceTimeoutMs > 0) {
-            accepted =
-                newExecutionsBackpressureSemaphore.tryAcquire(
-                    acceptanceTimeoutMs, TimeUnit.MILLISECONDS);
-          } else {
-            accepted = newExecutionsBackpressureSemaphore.tryAcquire();
-          }
-          if (!accepted) {
-            log.warn(
-                "LocalActivity queue is full and submitting timed out for activity {} with acceptanceTimeoutMs: {}",
-                activityTask.getActivityId(),
-                acceptanceTimeoutMs);
+          try {
+            TimeLimiter timeLimiter = SimpleTimeLimiter.create(timeoutThreadPool);
+            permit =
+                timeLimiter.callWithTimeout(
+                    () -> slotSupplier.reserveSlot(reservationCtx),
```

Comment:
I think this is showing `slotSupplier` needs to have a method to try to reserve a slot with a timeout. A lot of slot suppliers can be more efficient by taking the timeout directly without needing to spin up a bunch of threads .

Reply:
I don't like really like how that asks more of implementers, but I do get why the additional thread usage would be potentially problematic with tons of LAs. It'd be nice if there was some alternative to forcing it on the user but, I don't really see an obvious one.

The other option that seems to come to mind is to change this to a call to `tryReserveSlot`, polled on some interval, and then one thread can handle polling for all LAs, but, that could potentially have some weird fairness side-effects.

Reply:
Your not really asking more, just something different. The current implementation ask for them to make their code interruption aware, which isn't always the default. It also provides no way for the implementer to know the deadline of the request. I wouldn't do polling for fairness reasons and it still doesn't give the interface all the information. It is very normal in Java for interfaces like this to have overloads for blocking get, non blocking get, blocking with timeout, see [lock](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/Lock.html) for example.

Reply:
OK I will add the timeout to the interface